File: show/src/show/patternMatch.java
package show;
	public class patternMatch {
		short len;
		short cost; 
		short flags;
		int pemaByPatternEnd; // first PEMA entry in this source position the
		// PMA array belongs to that has the same
		// pattern and end
		int pemaByChildPatternEnd; // first PEMA entry in this source position
		// that has a child pattern = pattern and
		// childEnd = end
		short descendantRelationships;
		short pattern;

		public patternMatch(LittleEndianDataInputStream rs) {
			len = rs.readShort();
			cost = rs.readShort();
			flags = rs.readShort();
			rs.readShort(); // must read this because C++ compiled has 32 bit alignment
			pemaByPatternEnd = rs.readInteger();
			pemaByChildPatternEnd = rs.readInteger();
			descendantRelationships = rs.readShort();
			pattern = rs.readShort();
		}
	}

File: Web/pyLPBackEnd/EclipseWorkspace/pyLPBackEnd/src/PatternMatch.py
"""PatternMatch.py - One PMA (pattern-match array) element, 20 bytes.

Overview:
    Unpacks len, cost, flags, two PEMA indexes (by pattern-end and
    child-pattern-end), descendantRelationships, and pattern id.
    Mirrors C++ cPatternMatch.

Pipeline position:
    Loaded as Source.pma[]; used to highlight winning patterns.
"""
import struct

class PatternMatch:
    # Unpack 4 shorts + 2 ints + 2 shorts (20 bytes). Fourth short ignored.
    def __init__(self, rs):
        self.len, self.cost, self.flags, _, self.pemaByPatternEnd, self.pemaByChildPatternEnd, \
        self.descendantRelationships, self.pattern = struct.unpack('<4h2i2h', rs.f.read(20))
